File: examples/kotlin/src/main/kotlin/com/example/app/shared/Logger.kt
package com.example.app.shared

import java.time.LocalDateTime

data class LogEntry(
    val level: LogLevel,
    val message: String,
    val context: Map<String, Any?>,
    val timestamp: LocalDateTime = LocalDateTime.now(),
)

enum class LogLevel { DEBUG, INFO, WARN, ERROR }

interface Logger {
    fun debug(message: String, context: Map<String, Any?> = emptyMap())
    fun info(message: String, context: Map<String, Any?> = emptyMap())
    fun warn(message: String, context: Map<String, Any?> = emptyMap())
    fun error(message: String, context: Map<String, Any?> = emptyMap())
}

class ConsoleLogger : Logger {
    override fun debug(message: String, context: Map<String, Any?>) = write(LogLevel.DEBUG, message, context)
    override fun info(message: String, context: Map<String, Any?>) = write(LogLevel.INFO, message, context)
    override fun warn(message: String, context: Map<String, Any?>) = write(LogLevel.WARN, message, context)
    override fun error(message: String, context: Map<String, Any?>) = write(LogLevel.ERROR, message, context)

    private fun write(level: LogLevel, message: String, context: Map<String, Any?>) {
        val entry = LogEntry(level, message, context)
        val line = kotlinx.serialization.json.Json.Default.toString(kotlinx.serialization.Serializable())
        when (level) {
            LogLevel.ERROR -> println("ERROR: $line")
            LogLevel.WARN -> println("WARN: $line")
            else -> println(line)
        }
    }
}

File: examples/rust/src/shared/logger.rs
use chrono::{Duration, Utc};
use serde::{Serialize, Deserialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LogEntry {
    pub level: LogLevel,
    pub message: String,
    pub context: std::collections::HashMap<String, String>,
    pub timestamp: chrono::DateTime<Utc>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum LogLevel {
    Debug,
    Info,
    Warn,
    Error,
}

pub trait Logger {
    fn debug(&self, message: &str, context: &[(&str, &str)]);
    fn info(&self, message: &str, context: &[(&str, &str)]);
    fn warn(&self, message: &str, context: &[(&str, &str)]);
    fn error(&self, message: &str, context: &[(&str, &str)]);
}

pub struct ConsoleLogger;

impl Logger for ConsoleLogger {
    fn debug(&self, message: &str, _context: &[(&str, &str)]) {
        println!("{{\"level\":\"debug\",\"message\":\"{}\"}}", message);
    }
    fn info(&self, message: &str, _context: &[(&str, &str)]) {
        println!("{{\"level\":\"info\",\"message\":\"{}\"}}", message);
    }
    fn warn(&self, message: &str, _context: &[(&str, &str)]) {
        eprintln!("{{\"level\":\"warn\",\"message\":\"{}\"}}", message);
    }
    fn error(&self, message: &str, _context: &[(&str, &str)]) {
        eprintln!("{{\"level\":\"error\",\"message\":\"{}\"}}", message);
    }
}
